src/mam/mod_mam_rdbms_prefs.erl

```diff
@@ -160,157 +190,85 @@ set_prefs(_Result, Host, UserID, _ArcJID, DefaultMode, AlwaysJIDs, NeverJIDs) ->
             -> mod_mam:preference().
 get_prefs({GlobalDefaultMode, _, _}, Host, UserID, _ArcJID) ->
-    SUserID = escape_integer(UserID),
-    {selected, Rows} =
-    mod_mam_utils:success_sql_query(
-      Host,
-      ["SELECT remote_jid, behaviour "
-       "FROM mam_config "
-       "WHERE user_id=", use_escaped_integer(SUserID)]),
+    {selected, Rows} = mongoose_rdbms:execute(Host, mam_prefs_select, [UserID]),
     decode_prefs_rows(Rows, GlobalDefaultMode, [], []).
 
-
 -spec remove_archive(mongoose_acc:t(), jid:server(), mod_mam:archive_id(), jid:jid()) ->
     mongoose_acc:t().
 remove_archive(Acc, Host, UserID, _ArcJID) ->
     remove_archive(Host, UserID),
     Acc.
 
 remove_archive(Host, UserID) ->
-    SUserID = escape_integer(UserID),
     {updated, _} =
-    mod_mam_utils:success_sql_query(
-      Host, ["DELETE FROM mam_config WHERE user_id=", use_escaped_integer(SUserID)]).
+        mongoose_rdbms:execute(Host, mam_prefs_delete, [UserID]).
 
 -spec query_behaviour(jid:server(),
-                      SUserID :: mongoose_rdbms:escaped_integer(),
-                      SRemLJID :: mongoose_rdbms:escaped_string(),
-                      SRemLBareJID :: mongoose_rdbms:escaped_string(),
-                      CheckBare :: boolean()
+                      UserID :: non_neg_integer(),
+                      BRemLJID :: binary(),
+                      BRemLBareJID :: binary()
         ) -> any().
-query_behaviour(Host, SUserID, SRemLJID, SRemLBareJID, CheckBare) ->
-    Result =
-    mod_mam_utils:success_sql_query(
-      Host,
-      ["SELECT remote_jid, behaviour "
-       "FROM mam_config "
-       "WHERE user_id=", use_escaped_integer(SUserID), " "
-         "AND (remote_jid='' OR remote_jid=", use_escaped_string(SRemLJID),
-               case CheckBare of
-                    false ->
-                        "";
-                    true ->
-                       [" OR remote_jid=", use_escaped_string(SRemLBareJID)]
-               end,
-         ")"]),
-    ?LOG_DEBUG(#{what => mam_query_behaviour_result,
-                 user_id => SUserID, result => Result}),
-    Result.
+query_behaviour(Host, UserID, BRemLJID, BRemLJID) ->
+    mongoose_rdbms:execute(Host, mam_prefs_select_behaviour,
+                           [UserID, BRemLJID]); %% check just bare jid
+query_behaviour(Host, UserID, BRemLJID, BRemLBareJID) ->
+    mongoose_rdbms:execute(Host, mam_prefs_select_behaviour2,
+                           [UserID, BRemLJID, BRemLBareJID]).
 
 %% ----------------------------------------------------------------------
 %% Helpers
 
--spec encode_behaviour(always | never | roster) -> string().
-encode_behaviour(roster) -> "R";
-encode_behaviour(always) -> "A";
-encode_behaviour(never)  -> "N".
+%% Possible error with mysql
+%% Reason "Deadlock found when trying to get lock; try restarting transaction"
+%% triggered mod_mam_utils:error_on_sql_error
+run_transaction_or_retry_on_abort(Host, F, UserID, Retries) ->
```

Comment:
1. Why is this a problem only for `prefs`? It seems like it should be generic and occur somewhere else as well - and should be handled in `mongoose_rdbms`.
2. Does this deadlock happen during regular tests? If so, I think the problem would escalate during load tests? Maybe our transaction logic/isolation level is broken? I would investigate this as maybe it's just broken tests, e.g. setting prefs in parallel for the same user.
3. How did you test it?

Reply:
1. yep, there is generic version in mongoose_rdbms, but it does not has delay (because it's inside a worker, we don't want to wait in the DB worker).
2. regular. Waaaay too often during tests, rarely IRL.
3. Big tests sometimes fail without it (and actually it prints sometimes this warning in log).

Reply:
we probably can use just a builtin